### PR TITLE
Standard Improvements

### PR DESCRIPTION
**Pull request by PSchmiedmayer**
# Standard Improvements

## :recycle: Current situation & Problem
- Standards are currently challenging to modify and develop due to a large number of constraints and a rigid data source and provider enforcement
- Components all require a defined `ComponentStandard`, making it complicated for beginners to understand `Components` due to their generic syntax.
- All `Components` have to agree on a single `Standard` instance; this becomes increasingly challenging in larger projects and when a `Standard` is not defined, requiring a large infrastructure around `Adapter`s, data storage providers, and data sources.
- It is challenging to generally reference `Component`s, e.g., in reusable views where the `Standard` might not be defined. 

## :bulb: Proposed solution
- We propose to restructure the `Standard` definition to be optional for `Component`s. It is only limited to those who use the `@StandardActor` property wrapper.
- As a consequence, we suggest to removes the adapter, data storage provider, and data source protocols to simplify the code base. The functionality of how incoming data is handled is very application specific and should be handled in custom implementations of a `Standard`. 
   - Standards like the FHIR standard can be transformed into a component that makes it easier to locally deal with FHIR data.
   - Components like the Firebase component should focus on enabling an easy-to-use API surface for interacting with Firebase services; a user will have to write a small amount of well-abstracted code to link incoming data to current data storage providers like Firebase. This will allow more customization of where data is uploaded and how it is stored. 

## :gear: Release Notes
- Removed the `ComponentStandard` `associatedType` of a `Component`. The type alias or generic type constraints in current components can be transformed, as shown in the example below.
- Removes the `Adapter` and `DataStorageProvider`. Existing modules are encouraged to transform their usage of these APIs into `Standard` constraints, as shown in the example below:

#### 1. Standard Constraint

Define a standard constraint that is required by your component.
The constraint protocol **must** conform to the `Standard` protocol.
```swift
protocol ExampleConstraint: Standard {
    // ...
}
```

#### 2. Use the `@StandardActor` with the constraint

Use the constraint in your component to access the `Standard` instance that conforms to the protocol.
```swift
class ExampleComponent: Component {
    @StandardActor var standard: any ExampleConstraint
    
    // ...
}
```

#### 3. Standard Definition 

Ensure that you define an appropriate standard in your configuration in your `SpeziAppDelegate` subclass:
```swift
var configuration: Configuration {
    Configuration(standard: ExampleStandard()) {
        // ...
    }
}
```

#### 4. Standard Conformance 

The `Standard` defined in the `Configuration` must conform to all constraints defined by `Components` using their `@StandardActor`s or you need to write an extension to an existing `Standard` that you use to implement the conformance.

```swift
actor ExampleStandard: Standard, ExampleConstraint {
    // ...
}
```

### Loss in Type-Safety

Unfortunately, the new API can not directly enforce that a `Standard` has to conform to the protocols that are defined in the `Component`s `@StandardActors`. We currently check the correct conformance using several `guard` and `preconditionFailure` calls when injecting the `Standard` in the `@StandardActor` property wrappers: [StandardPropertyWrapper.swift](https://github.com/StanfordSpezi/Spezi/blob/feature/standardImprovement/Sources/Spezi/Standard/StandardPropertyWrapper.swift).
We can also not enforce that the passed-in protocol to `@StandardActor` conforms to `Standard` as only concrete types can be used for generic constraints resulting in an `Type 'any ...' cannot conform to 'Standard'`:
```swift
// 1. Adapt `_StandardPropertyWrapper`:
public class _StandardPropertyWrapper<Constraint: Standard>: AnyStandardPropertyWrapper { /* ... */ }

// 2. Example of error
protocol ExampleConstraint: Standard {
    // ...
}

class ExampleComponent: Component {
    @StandardActor var standard: any ExampleConstraint // Error: Type 'any ExampleConstraint' cannot conform to 'Standard'
    
    // ...
}
```

An alternative considered was using the current `CompoenentStandard` `associatedtype` to provide some protocol-based constraints on the `Standard` that could, e.g., be enforced in the `ComponentBuilder`. Unfortunately, Swift does not allow us to express this constraint based on protocol conformance, as discussed in the following Swift Forums thread: https://forums.swift.org/t/generic-where-fails-for-protocol-types/63729.

An alternative that would be worth exploring in the future would be the usage of two Swift macros to enforce the following constraints.
- `@StandardActor` macro that enforces that the based in type conforms to `Standard`, reducing the `preconditionFailure` check in the current initializer of the `_StandardPropertyWrapper` and moving this to a compiler time error
- `#Configuration` macro that replaces the current `Configuration(standard: Standard) { ... }` initializer + result builder setup. The macro can 
   1. Collect all `Components` in the result builder that have a `@StandardActor` in their body (extracted using reflection) and collect all type constraints enforced using these property wrappers
   3. Check this against the Swift Syntax information provided for the Standard (by checking all protocols it conforms to) and subsequently compare them to the list obtained in the first step.

Both approaches would require additional investigation into the Swift Macro system.

### Related Issues
- Addresses #65

### Testing
- The tests were adapted, and any code related to adapters, data storage providers, and data sources was removed
- XCTSpezi no longer contains any code as a default standard has decreased in complexity and, most often, is no longer needed. The framework target should be extended in the future to make testing easier by automating some of the testing functionality.

### Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by PSchmiedmayer** [comment]
I will add some additional tests to the PR to check that the preconditions are correctly enforced when using a `Standard` constraint; apart from that, the PR would be ready for a first review. This will also address the diff code coverage.

In combination with #72 that will be a huge step forward for the reusability across components, a lot of the idea here grew from the discussions with @Supereg about some of the limitations we currently have 🎉

Would love to get your thought about the tradeoffs defined in the PR description and the thoughts on the loss of type-safety compared to a more flexible system @vishnuravi @Supereg @philippzagar 🚀

**Comment by vishnuravi** [comment]
This PR is a great improvement to the framework - it addresses a lot of the feedback we've had from users who have had trouble understanding how to get started due to the current complexity of the `Standard`. Great work @PSchmiedmayer and @philippzagar! The PR looks good to merge from my perspective!

**Comment by PSchmiedmayer** [comment]
Thank you for the feedback @vishnuravi! The plan would be to merge the PR some time today, then tag a new major version release and the bake this a bit by updating the different modules step by step over the next weeks while #72 will also be merged and incorporated in some modules 🚀

**Comment by vishnuravi** [comment]
> Thank you for the feedback @vishnuravi! The plan would be to merge the PR some time today, then tag a new major version release and the bake this a bit by updating the different modules step by step over the next weeks while #72 will also be merged and incorporated in some modules 🚀

That plan sounds good to me! We will also need to update the documentation to reflect these changes, though I think that can be done in a different PR.

**Comment by vishnuravi** [comment]
> I think loosing type-safety is manageable as this is probably the best alternative. At least I couldn't think of a better solution. Thanks for the detailed PR description discussing all the alternatives you though of 🥳

I agree, it is a worthwhile tradeoff here.

File: Sources/Spezi/Standard/StandardPropertyWrapper.swift [hunk 1]
```diff
@@ -33,10 +33,35 @@ public class _StandardPropertyWrapper<S: Standard> {
...
+                ```
+                
+                ... and that your standard conforms to \(String(describing: Constraint.self)):
+                
+                ```swift
+                actor \(String(describing: S.self)): Standard, \(String(describing: Constraint.self)) {
+                    // ...
+                }
+                ```
+                """
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply Sources/Spezi/Standard/StandardPropertyWrapper.swift:61]: Thank you 🚀

File: Sources/Spezi/Configuration/Configuration.swift [hunk 2]
```diff
@@ -53,13 +53,32 @@ public struct Configuration {
     
 
     /// A ``Configuration`` defines the ``Standard`` and ``Component``s that are used in a Spezi project.
+    ///
+    /// If no ``Standard`` is provided, Spezi injects a default empty standard instance only conforming to ``Standard``.
```
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply Sources/Spezi/Configuration/Configuration.swift:57]: Agree, I will add some more documentation around there.

**Comment by vishnuravi** [comment]
> Thank you @vishnuravi & @Supereg for the feedback. I plan to create some docs as part of this PR before I merge it today and step on the flight.
> 
> I will also add some additional issues about some ideas around documentation that would be great to have beyond what we will have in the current documentation. I have some general ideas to e.g. add an article on how to get started with running your first Spezi App, but that might even be better located in the Template App. Will see where I add these Ideas 👍

That sounds great! I agree that the "Getting Started" guide should be in the Template Application. There is a PR here in progress for that which we can use: https://github.com/StanfordSpezi/SpeziTemplateApplication/pull/19

**Comment by philippzagar** [comment]
Great job @PSchmiedmayer!🚀
I didn't have the chance yet to do an in-depth review of the PR but from a high-level perspective and the listed tradeoffs in the PR description, the changes look very senseful! I'll give feedback next week.

**Comment by philippzagar** [comment]
Great work @PSchmiedmayer with the simplification of the `Component` usage in Spezi! I especially liked your listed tradeoffs outlining the benefits and drawbacks of the proposed approaches! 🚀
Also thanks a lot for going the extra mile by writing nice documentation and migration guidelines, very beneficial for Spezi users!

I think the loss in type safety is acceptable here (as it definitely reduces the complexity surrounding Spezi's `Standard`s), especially as we might be able to use the new Swift Macro language feature to compensate for that. We should definitely investigate your outlined approaches soon!

**Pull request merged**